instant-segment/examples/merge.rs

```diff
@@ -0,0 +1,246 @@
+            let fname = format!("data/cache/eng-1-{:05}-{:05}.txt", part, UNIGRAM_PARTS);
+            let f = File::open(&fname).unwrap();
+            let mut reader = BufReader::with_capacity(4 * 1024 * 1024, f);
+
+            let mut ln = String::new();
+            let mut unigrams = HashMap::with_capacity(8 * 1024);
+            loop {
+                ln.clear();
+                match reader.read_line(&mut ln) {
+                    Ok(0) => break,
+                    Err(e) => {
+                        eprintln!("error: {:?}", e);
+                        break;
+                    }
+                    _ => {}
+                }
+
+                let mut iter = ln.trim().split('\t');
+                let word = match iter.next() {
+                    Some(word) => word,
+                    _ => continue,
+                };
+
+                let word = match normalize(word, word_list) {
+                    Some(word) => word,
+                    _ => continue,
+                };
+
+                let mut matches = 0;
+                for year_data in iter {
+                    let mut parts = year_data.split(',');
+                    if parts.next().unwrap() < START_YEAR {
+                        continue;
+                    }
+                    matches += usize::from_str(parts.next().unwrap()).unwrap();
+                }
+
+                *unigrams.entry(word).or_default() += matches;
+            }
+
+            eprintln!("extracted {} unigrams from part {}", unigrams.len(), part);
+            unigrams
+        })
+        .reduce(HashMap::<SmartString, usize>::new, |mut left, right| {
+            for (k, v) in right.into_iter() {
+                *left.entry(k).or_default() += v;
+            }
+            left
+        });
+
+    let mut unigrams = unigrams.into_iter().collect::<Vec<_>>();
+    unigrams.sort_by_key(|(_, freq)| Reverse(*freq));
+    let f = File::create("data/en-unigrams.txt").unwrap();
+    let mut writer = BufWriter::with_capacity(4 * 1024 * 1024, f);
+    for (i, (word, freq)) in unigrams.into_iter().enumerate() {
+        if i == MAX_UNIGRAMS {
+            break;
+        }
+
+        writeln!(writer, "{}\t{}", word, freq).unwrap();
+    }
+}
+
+/// Read the word list and gather it up into a hash set for easy lookups
+///
+/// We use this to filter crappy words out of the (pretty noisy) ngram data.
+/// Considering the way we want to [`normalize()`], we'll filter for
+/// only-letter contents but keep any uppercase characters intact.
+fn read_word_list() -> HashSet<SmartString> {
+    const AVERAGE_WORD_LIST_LINE_LEN: usize = 9;
+
+    let f = File::open("data/cache/eng-wordlist.txt").unwrap();
+    let size = f.metadata().unwrap().len() as usize;
+    let mut reader = BufReader::with_capacity(4 * 1024 * 1024, f);
+
+    eprintln!("read word list...");
+    let mut word_list = HashSet::with_capacity(size / AVERAGE_WORD_LIST_LINE_LEN);
+    let mut ln = String::new();
```

Comment:
Maybe add a comment here with what a line looks like in the data? I think that would make the parsing logic easier to grok.

Reply:
Oh yeah, good idea. Added that.